utils/http.go

```diff
@@ -94,3 +96,102 @@ func buildAccessRecords(repo string, actions ...string) []auth.Access {
 	}
 	return requiredAccess
 }
+
+// CacheControlConfig is an interface for something that knows how to set cache
+// control headers
+type CacheControlConfig interface {
+	// SetHeaders will actually set the cache control headers on a Headers object
+	SetHeaders(headers http.Header)
+}
+
+// NewCacheControlConfig returns CacheControlConfig interface for either setting
+// cache control or disabling cache control entirely
+func NewCacheControlConfig(maxAgeInSeconds int, mustRevalidate bool) CacheControlConfig {
+	if maxAgeInSeconds > 0 {
+		return PublicCacheControl{MustReValidate: mustRevalidate, MaxAgeInSeconds: maxAgeInSeconds}
+	}
+	return NoCacheControl{}
+}
+
+// PublicCacheControl is a set of options that we will set to enable cache control
+type PublicCacheControl struct {
+	MustReValidate  bool
+	MaxAgeInSeconds int
+}
+
+// SetHeaders sets the public headers with an optional must-revalidate header
+func (p PublicCacheControl) SetHeaders(headers http.Header) {
+	cacheControlValue := fmt.Sprintf("public, max-age=%v, s-maxage=%v",
+		p.MaxAgeInSeconds, p.MaxAgeInSeconds)
+
+	if p.MustReValidate {
+		cacheControlValue = fmt.Sprintf("%s, must-revalidate", cacheControlValue)
+	}
+	headers.Set("Cache-Control", cacheControlValue)
+	// delete the Pragma directive, because the only valid value in HTTP is
+	// "no-cache"
+	headers.Del("Pragma")
+	if headers.Get("Last-Modified") == "" {
+		SetLastModifiedHeader(headers, time.Time{})
+	}
+}
+
+// NoCacheControl is an object which represents a directive to cache nothing
+type NoCacheControl struct{}
+
+// SetHeaders sets the public headers cache-control headers and pragma to no-cache
+func (n NoCacheControl) SetHeaders(headers http.Header) {
+	headers.Set("Cache-Control", "max-age=0, no-cache, no-store")
+	headers.Set("Pragma", "no-cache")
+}
+
+// cacheControlResponseWriter wraps an existing response writer, and if Write is
+// called, will try to set the cache control headers if it can
+type cacheControlResponseWriter struct {
+	http.ResponseWriter
+	config     CacheControlConfig
+	statusCode int
+}
+
+// WriteHeader stores the header before writing it, so we can tell if it's been set
+// to a non-200 status code
+func (c *cacheControlResponseWriter) WriteHeader(statusCode int) {
+	c.statusCode = statusCode
+	c.ResponseWriter.WriteHeader(statusCode)
+}
+
+// Write will set the cache headers if they haven't already been set and if the status
+// code has either not been set or
```

Comment:
Seems like an incomplete comment, should it be completed with `...or is not 200`?